src/compiler/binder.ts

```diff
@@ -3340,6 +3341,9 @@ function createBinder(): (file: SourceFile, options: CompilerOptions) => void {
                 }
             });
         }
+        if (isPrototypeProperty && namespaceSymbol && namespaceSymbol.valueDeclaration && isAssignmentDeclaration(namespaceSymbol.valueDeclaration)) {
+            containerIsClass = false;
+        }
         if (containerIsClass && namespaceSymbol && namespaceSymbol.valueDeclaration) {

```

Comment:
I think you can combine the checks from the above as well and just do an early return in the "bad" case.

Reply:
I'm not sure I understand this. Do you mean I should add it as an `else if` ?

Reply:
Something like that - you're setting `containerIsClass` to `false`, which makes you indirectly skip the next check; maybe that's fine, but you also have to repeat the same checks for `namespaceSymbol`. So I would write something like

```ts
if (namespaceSymbol?.valueDeclaration) {
    // We may have needed to make an assumption that the container symbol is a class
    // due to an assignment like `Foo.prototype.method = ...`. However,
    // if we never found a valueDeclaration that's suited as an old-style class,
    // we should bail out here.
    if (isPrototypeProperty && isAssignmentDeclaration(namespaceSymbol.valueDeclaration)) {
        return namespaceSymbol;
    }

    if (containerIsClass) {
        addDeclarationToSymbol(namespaceSymbol, namespaceSymbol.valueDeclaration, SymbolFlags.Class);
    }
}
```

You can even simplify further, but I think this is fine.

Reply:
On that note - maybe for the bail-out case it would be better to say

> If the `valueDeclaration` is *not* a class or constructor function

instead of

> If the `valueDeclaration` is an assignment function

That way we can't end up with more accidental cases like this. But it's up to you.